utils/generate-command-code.py

```diff
@@ -68,6 +67,51 @@ def get_optional_desc_string(desc, field, force_uppercase=False):
     return ret.replace("\n", "\\n")
 
 
+def check_command_key_specs(cmd):
+    if not command.key_specs:
+        return
+
+    assert isinstance(command.key_specs, list)
+
+    for cmd_key_spec in command.key_specs:
+        if "flags" not in cmd_key_spec:
+            print("command: %s key_specs missing flags" % command.fullname())
+            return
+
+        if "NOT_KEY" in cmd_key_spec["flags"]:
+            # Like SUNSUBSCRIBE / SPUBLISH / SSUBSCRIBE
+            return
+
+    command_key_specs_index_set = set(range(len(command.key_specs)))
+    command_arg_key_specs_index_set = set()
+
+    # Collect key_spec used for each arg, including arg.subarg
+    for arg in command.args:
+        if arg.key_spec_index is not None:
+            assert isinstance(arg.key_spec_index, int)
+
+            if arg.key_spec_index not in command_key_specs_index_set:
+                print("command: %s arg: %s key_spec_index error" % (command.fullname(), arg.name))
+                return
+
+            command_arg_key_specs_index_set.add(arg.key_spec_index)
+
+        for sub_arg in arg.subargs:
+            if sub_arg.key_spec_index is not None:
+                assert isinstance(sub_arg.key_spec_index, int)
+
+                if sub_arg.key_spec_index not in command_key_specs_index_set:
+                    print("command: %s arg: %s key_spec_index error" % (command.fullname(), sub_arg.name))
+                    return
+
+                command_arg_key_specs_index_set.add(sub_arg.key_spec_index)
```

Comment:
shouldn't this be recursive? at least in theory?
i guess we don't have key arguments in deep nested args..

Reply:
> shouldn't this be recursive? at least in theory?

yes, it should, this needs to be changed... i thought we only have one level sub arg, but i was wrong

> i guess we don't have key arguments in deep nested args..

yes

Reply:
btw, should i change it to be recursive?

Reply:
it's not needed at the moment, but still it'l be more "correct" so i'd suggest to change it.